sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc

```diff
@@ -0,0 +1,651 @@
+
+IMPORTANT: In the current implementation, only the subgroup scope is supported
+
+When the group is a `sycl::sub_group`, a matrix is declared as follows:
+
+```c++
+joint_matrix<int8_t, tM, tN, matrix_use::matrix_a> tA(sg); 
+```
+
+
+## Matrix Operations and their Execution Scope
+We define three new functions needed to perform the main and common operations on matrices namely, load, store, and the actual multiply and add operation. This set of functions can be easily extended if the tensor hardware implements new features.
+
+The base pointer determines the starting address of the matrix to be loaded/stored. `layout` determines whether the data are being read/written in a row (`row_major`), column major (`column_major`) fashion, or if the data has already been transformed into VNNI format (`packed`). `stride` describes the number of elements between consecutive rows for row major and packed layout,  columns for column major layout. 
+
+Note that for getting maximum performance on Intel AMX and DPAS, prepacking data in the memory is necessary. If users did not specify the packed layouts, transforms done by the implementation will be slow due to extra scatter/gather operations. Hence, we expose the `packed` layout to the user to specify that A or B have already been VNNIed. The packed or VNNI layout is introduced in `VNNI layout` section below.
+	
+IMPORTANT: In the current AMX and DPAS implementation, the layout in the load of matrix B must be `packed_b` or `row_major`. Automatic VNNI transform is supported. The layout in the load of matrices A and C must be `row_major`, and the layout in the store of matrix C must also be `row_major`.
+
+Since the matrix functions are group operations (as defined in Section 4.17.3 of the SYCL specification), the matrix API has to be accessed by all the work-items in the group in a convergent control flow. The `Group` template argument can be a work-group or a subgroup. These functions will be called once by each work item in the group.
+
+To be aligned with the SYCL 2020 group algorithms, an additional group argument is added to the matrix operations to designate that these functions are collective operations. The {dpcpp} syntax is the following: 
+
+IMPORTANT: In the current implementation, only the subgroup scope is supported.  
+
+#### Load 
+```c++
+namespace sycl::ext::oneapi::experimental::matrix {
+  template <typename Group, typename T, size_t NumRows, size_t NumCols,
+          matrix_use U, matrix_layout L,
+          access::address_space Space>
+  void joint_matrix_load(Group sg, joint_matrix<T, NumRows, NumCols, U, L, Group> &res,
+		    multi_ptr<T, Space> src, size_t stride, matrix_layout memL);
```

Comment:
Can you clarify the distinction between the `matrix_layout L` template parameter and `matrix_layout memL` parameter here?

Reply:
Specifically do you plan on keeping the `matrix_layout L` template parameter in `joint_matrix_load` and `joint_matrix_store`?
Perhaps you have only kept this in because you think it is necessary for compatibility with the CUDA backend, since I notice that in the latest implementation, https://github.com/intel/llvm/pull/5835, L is set as layout::unused by default?
In the CUDA implementation I took the approach of overloading `joint_matrix_load`: the first overload, https://github.com/intel/llvm/blob/e55e5f0bfe5175d41f61f4b912243a988104e2f7/sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp#L752, was intended to be compatible with the Intel backend, because `layout` is provided as a normal runtime parameter. This is used in the CUDA backend to load the accumulator matrices.
The second overload, https://github.com/intel/llvm/blob/e55e5f0bfe5175d41f61f4b912243a988104e2f7/sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp#L780, was intended to only be used in the CUDA and HIP AMD backends, because it is required for the A/B "multiplicand" matrices for which the layout is required at compile time. In this overload there is no runtime `layout` parameter; instead the layouts of the matrices are inferred from the `joint_matrix` arguments.

In the Intel backend can the layouts of all matrices be specified at runtime? Or is it the case that, as with the CUDA backend, some of them require that the layout is known at compile time?

Reply:
Same here, we remove L from the template arguments of store and replace it with "unused" in joint matrix.